### ISSUE TITLE
softmax warning, putting tensors on the active device

### ISSUE THREAD

**Issue by AngledLuffa** [v1xerunt/PyTorch-Dynamic_LSTM#3]
There is a warning produced by pytorch for the usage of softmax.  I could clear that up by doing this:

```
         self.h2h = nn.Linear(self.n_hidden, 4 * self.n_hidden)
         self.output = nn.Linear(self.n_hidden, self.n_output)
         self.sigmoid = nn.Sigmoid()
-        self.softmax = nn.Softmax()
+        self.softmax = nn.Softmax(dim=-1)
         self.dropout = nn.Dropout(p=self.dropout)
         self.tanh = nn.Tanh()
```

If an input tensor is passed in which is not on the default device, especially if the DynamicLSTM has been moved to GPU, there are tensor operations which try to operate across devices (causing a failure).  The following changes helped with that, although there may be better ways of doing it:

```
         if agent_prob is None:
             agent_prob = []  # Probabilities for agents
+        device = input.device
         # Hidden state for lstm
-        cur_h = Variable(torch.zeros(batch_size, self.n_hidden))
+        cur_h = Variable(torch.zeros(batch_size, self.n_hidden, device=device))
         # Cell memory for lstm
-        cur_c = Variable(torch.zeros(batch_size, self.n_hidden))
+        cur_c = Variable(torch.zeros(batch_size, self.n_hidden, device=device))
         full_c = []  # Cell memory list for lstm
         full_h = []  # Hidden state list for lstm


                     torch.cat((input[:, 0, :], cur_h), 1), cur_time, agent_action, agent_prob)
-                observed_c = torch.zeros_like(cur_c, dtype=torch.float32).view(-1).repeat(
+                observed_c = torch.zeros_like(cur_c, dtype=torch.float32, device=device).view(-1).repeat(
                     self.n_actions).view(self.n_actions, batch_size, self.n_hidden)
-                observed_h = torch.zeros_like(cur_h, dtype=torch.float32).view(-1).repeat(
+                observed_h = torch.zeros_like(cur_h, dtype=torch.float32, device=device).view(-1).repeat(
                     self.n_actions).view(self.n_actions, batch_size, self.n_hidden)


                 observation = torch.cat((input[:, cur_time, :], cur_h), 1)
                 actions = self.choose_action(observation, cur_time, agent_action, agent_prob)
                 coord = torch.cat((actions.int(), torch.arange(
-                    batch_size, dtype=torch.int).unsqueeze(-1)), 1)
+                    batch_size, dtype=torch.int, device=device).unsqueeze(-1)), 1)
```
